### PR TITLE
[WIP] Static analysis with PHPStan

### PR DESCRIPTION
**Pull request by ondrejmirtes**
Continuing the work from #9898. This is PR to integrate [PHPStan](https://github.com/phpstan/phpstan) into the CI build of CakePHP.

It's a work in progress since there are quite a lot of errors in the build log. `@property` annotations are already supported by PHPStan, but it still found some undefined properties.

Even after merging this PR, I expect to continue to work with you on increasing the level of strictness of the analysis and also on updating to newer versions of PHPStan :)

I'd like one of the maintainers to go through the error log and fix obvious errors in the code. We can then work together on the other, more questionable, errors.

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Also, the mergable version of this PR will use a stable version of PHPStan. I'm waiting with the release in case we find something to fix in it.

File: .travis.yml
Comment (ADmad) [inline .travis.yml]: This will make phpstan check run along with the phpcs check. It would be better to have the phpstan check in a job of it's own so that it can be added to "allow_failures" for now. It will be a while before we can make all required changes to code to make it pass.

**Review by ADmad — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (ondrejmirtes) [inline .travis.yml]: Should be better now.

File: .travis.yml
Comment (dereuromark) [inline .travis.yml]: Do we need both, or should we just have 7.1 here (and also in allowed failure list then). The result should be the same pretty much either way, right?

Comment (ondrejmirtes) [inline .travis.yml]: It will differ in case that something in PHP changed because I use reflection of the current version.

**Comment by ADmad** [comment]
> Also, the mergable version of this PR will use a stable version of PHPStan. I'm waiting with the release in case we find something to fix in it.

Using the master branch for a while is not a problem, at least until the phpstan job is under "allow_failures". That way you too have a sizeable codebase to test against when making updates to phpstan :smile:

File: .travis.yml
Comment (ondrejmirtes) [inline .travis.yml]: I think it's a good idea to run both.

Comment (ADmad) [inline .travis.yml]: Yes testing against PHP 7 and 7.1 doesn't seem necessary especially when current CakePHP version support PHP 5.5+ and we don't use any PHP 7 features.

Comment (dereuromark) [inline .travis.yml]: I am just a bit worried about the average build time then as it will further increase with each specific run. But I guess for the time being it doesnt hurt too much and those phpstan builds are probably rather fast.

File: .travis.yml
Comment (dereuromark) [inline .travis.yml]: For a failure list to work you also need to include the php version each time afaik.

Comment (ondrejmirtes) [inline .travis.yml]: I'm not sure, we'll see :)

Comment (ADmad) [inline .travis.yml]: @ondrejmirtes https://travis-ci.org/cakephp/cakephp/builds/187472461  so @dereuromark is right.

**Comment by ADmad** [comment]
@ondrejmirtes Houston we have a problem https://travis-ci.org/cakephp/cakephp/jobs/187472483#L628

File: .travis.yml
Comment (dereuromark) [inline .travis.yml]: Let's get one (7.0 then maybe) to work first, then lets think about adding more.

**Comment by ondrejmirtes** [comment]
@ADmad That's actually a problem with the code of CakePHP :) You cannot use `Exception` because `Cake\Database\Exception` already exists. So when you reference an `Exception` in the Connection class, PHP does not know which exception you mean. I will add a commit to this PR fixing that.

**Comment by ADmad** [comment]
> That's actually a problem with the code of CakePHP :)

Well the code works perfect fine in it's current state :)

**Comment by ondrejmirtes** [comment]
@ADmad Only thanks to a coincidence. If you `require` the Exception first and the Connection second, you will end up with this error.

**Comment by dereuromark** [comment]
Looking pretty good already

A few simple ones like

    Call to method Cake\ORM\Association\BelongsToMany::_selectableoptions()                                    
         with incorrect case: _selectableOptions      

are relatively easy to fix, I can take care of those after merge, as well.

File: src/Database/Connection.php
Comment (dereuromark) [inline src/Database/Connection.php]: Why is this necessary again? It should be OK to not use FQCN inside code and always include them via use statements.

File: src/Http/Client/Auth/Digest.php
Comment (dereuromark) [inline src/Http/Client/Auth/Digest.php]: Not sure we can change the signature in a bugfix release on master, for reasons of BC.

File: src/View/Form/NullContext.php
Comment (dereuromark) [inline src/View/Form/NullContext.php]: Same here..

**Comment by dereuromark** [comment]
As for the Type changes, I am 1000% with you on those, but that was a constant pull and push here regarding some of the other core devs wanting to have those bad code blocks in there for theoretical future refactoring - see https://github.com/cakephp/cakephp/pull/9907
So it might be best for you not to open up that can of worms again ;)

File: src/Auth/DigestAuthenticate.php
Comment (ADmad) [inline src/Auth/DigestAuthenticate.php]: Seeing this change I remember I have already fixed this and few other things on the `3.next` branch. It would be best if this PR was targeted to `3.next` instead of `master`.

**Comment by ADmad** [comment]
@ondrejmirtes Would you mind opening a new pull request against the `3.next` with just modifications to `.travis.yml` and the `phpstan.neon` file?  We can then review the report and make code changes or add ignores as required.

**Comment by dereuromark** [comment]
@ADmad It does contain some fixes that could be valuable for the current master. What do you think, can be still merge those?

**Comment by ADmad** [comment]
Given that 3.4 is round the corner I don't think it's worth the effort.

File: src/Database/Connection.php
Comment (ondrejmirtes) [inline src/Database/Connection.php]: It's basically this problem:

https://gist.github.com/cvuorinen/e7a19101d40c9fa0a959
https://3v4l.org/cphKu

File: src/Http/Client/Auth/Digest.php
Comment (ondrejmirtes) [inline src/Http/Client/Auth/Digest.php]: I think this change is fine. The second parameter was optional and not used anyway, so it cannot break any code.

File: src/View/Form/NullContext.php
Comment (ondrejmirtes) [inline src/View/Form/NullContext.php]: The problem is that the parameter may be unused because of two reasons:

* It's really not supposed to be used and should be removed.
* It's forgotten to be used and the code should be fixed. Like in this code sample:

```
private $foo;

private $bar;

public function __construct($foo, $bar)
{
    $this->foo = $foo;
}
```

This check actually found a few of these situations in our codebase so I'm convinced it should be part of the default level 0.

**Comment by ondrejmirtes** [comment]
OK, I don't know your Git workflow so I assumed I should open the PR on master. I will make a new one and we'll see how it builds :)

File: src/Http/Client/Auth/Digest.php
Comment (ondrejmirtes) [inline src/Http/Client/Auth/Digest.php]: I originally removed the Response parameter from SessionStorage but returned it later because that change could actually break something.

**Comment by ADmad** [comment]
> OK, I don't know your Git workflow so I assumed I should open the PR on master.

No problem, my fault for noticing the target branch sooner.

**Comment by ondrejmirtes** [comment]
Continuing in #9943 :)

**Pull request closed without merging**
